gh-123523: Rework typing documentation for generators and coroutines, and link to it from `collections.abc` docs (#123544)
python/cpython: Doc/library/collections.abc.rst

.. versionadded:: 3.9
   These abstract classes now support ``[]``. See :ref:`types-genericalias`

>    and :pep:`585`. They, however, do not inherit from :class:`~typing.Generic` explicitly.

Reviewer: Is this added note necessary? The existing `types-genericalias` link takes you to https://docs.python.org/3/library/stdtypes.html#types-genericalias, which I think makes it clear that they don't explicitly inherit from generics

Author: Indeed it's a remainder from a now reverted attempt to list generic arguments there - it can be removed now.

----------------------------------------------------------

> .. class:: Container[T_co]

Reviewer: Last time I checked, using PEP-695 syntax broke Sphinx's rendering for the `.. class::` directive (https://github.com/python/cpython/pull/104642#discussion_r1200799012), at least with the Sphinx version we have pinned for our docs. Has that changed since?

Author: At least it looks fine in [preview](https://cpython-previews--123544.org.readthedocs.build/en/123544/library/collections.abc.html#collections.abc.Coroutine) - what didn't work before, exactly?

Reviewer: > what didn't work before, exactly?

I can't really give a better explanation than the screenshots I shared in https://github.com/python/cpython/pull/104642#discussion_r1200799012, I don't think (click the "details" drop-downs in that comment to see them)

Reviewer: According to @AA-Turner in https://github.com/python/cpython/pull/104642#discussion_r1202366559, we shouldn't start using PEP-695 syntax in our docs until the `needs_sphinx` value in our `conf.py` is bumped to a version of Sphinx that supports rendering PEP-695 class declarations. `needs_sphinx` is currently at 6.2.1: https://github.com/python/cpython/blob/34ddb64d088dd7ccc321f6103d23153256caa5d4/Doc/conf.py#L87

And Sphinx added support for PEP-695 class declarations in v7.1.0 according to [their changelog](https://www.sphinx-doc.org/en/master/changes/7.1.html#features-added)

Author: Ough, this is really unfortunate - we need to list generic args somehow here. These classes don't inherit from `Generic` - would it be fine to pretend they do?

Reviewer: I'd prefer not to... I think this is another argument in favour of the solution I proposed in https://github.com/python/cpython/pull/123544#discussion_r1739932413 -- keep in-depth discussion of type hints in the `typing`-module docs (but in a new section rather than in the references for the deprecated symbols like `typing.Generator` etc.), and link to that in-depth discussion from the `collections.abc` docs

Author: I already started moving generators and coroutines explanations to such a section, but it doesn't remove the need to provide type arguments right there. I do definitely remember all such basic usage guidelines, but can go to that documentation to check whether it's `Generator[YieldT, ReturnT, SendT]` or `Generator[YieldT, SendT, ReturnT]`. I'd rather not scroll through wall of explanatory text in such case, linking to a howto is less desired for focused search - generic args are critical part of "class signature", they shouldn't be buried so deep.

Author: Hm, `Generic` [doesn't have](https://github.com/python/typing/issues/449) a custom metaclass any longer, so maybe it's a good compromise to just pretend these classes inherit from `Generic`?

Reviewer: > I'd rather not scroll through wall of explanatory text in such case, linking to a howto is less desired for focused search - generic args are critical part of "class signature", they shouldn't be buried so deep.

Makes sense. Another consideration is what these docs will look like to people less familiar with typing syntax, though -- we don't want to make the `collections.abc` docs impenetrable for those readers. I love typing as much as you do, but it's still an "optional add-on" to the language rather than something we want to enforce for all users... the declaration for `Generator`, for example, becomes quite a lot more complicated with this PR due to the three type parameters.

I worry that making this kind of change to the class declaration in the `collections.abc` docs could be quite controversial; it should possibly be discussed on Discourse first. Could we do it in a separate PR, and keep this PR focussed on adding the new section in the `typing` docs, and linking to that section from the `collections.abc` docs?

Author: Honestly, I'm becoming less and less of python's typing fan since the awkward PEP695 for a similar reason: too much attention and growing mindset of "sane python code should be typed". 

However, I agree that this might add more cognitive load than desired to these declaration. The problem of displaying generic args should be solved somehow, I'll open a separate ticket and revert this change here. As a quick suggestion: would you be more likely to accept adding explicit "type arguments" notice in every `collections.abc` class block? 

Something like this:

```rst
.. class:: Mapping
           MutableMapping

   ABCs for read-only and mutable :term:`mappings <mapping>`.

   Type arguments:

   * ``Mapping[KT, VT_co](Collection[KT])``
   * ``MutableMapping[KT, VT](Mapping[KT, VT])``
```

Reviewer: > As a quick suggestion: would you be more likely to accept adding explicit "type arguments" notice in every `collections.abc` class block?

yeah, that looks much more appealing to me!

Reviewer 2: Alex is right that we need to support Sphinx 6.2 (this is checked by CI).

Reviewer 2: > ```rst
>    Type arguments:
> 
>    * ``Mapping[KT, VT_co](Collection[KT])``
>    * ``MutableMapping[KT, VT](Mapping[KT, VT])``
> ```

Seeing this, I'm slightly less enthusiastic, as it is a lot of syntactic complexity all at once. For a reader new to typing, or even experienced with using typing, but not with type theory, this is a lot to stomach. Worse, it's unclear where the reader can learn what the mysterious symbols `VT_co` or `KT` mean / stand for. Perhaps apocryphal, but should we be expecting readers to learn about type variance before realising that ``Mapping[str, str]`` can be used?

Ideally these symbols could be linked directly to individual explanations, but if not, could we add a footnote or reference to an explanation of the standard typevar names and their implications? 

A

Reviewer: > Ideally these symbols could be linked directly to individual explanations, but if not, could we add a footnote or reference to an explanation of the standard typevar names and their implications?

We're no longer making these changes in this PR, so let's open a new issue after this has landed and discuss these questions there 👍

Reviewer 2: Ah, I didn't realise, sorry!

A

Reviewer 3: > And Sphinx added support for [PEP-695](https://peps.python.org/695) class declarations in v7.1.0 according to [their changelog](https://www.sphinx-doc.org/en/master/changes/7.1.html#features-added)

Yes but I needed to fix a bug in LaTeX docs, so ideally it should be using Sphinx 8.1 at least (and not just 7.1.x). We didn't get a lot of feedback for this feature and that's probably why the bug went unnoticed after the feature was released (I implemented the feature after Jelle asked for it).

   ABC for classes that provide the :meth:`~object.__call__` method.

>    See :ref:`annotating-callables` for details on how to use

Reviewer 3: I don't know if it was already addressed (opening the comments one by one is a bit long) but for consistency I would say "Refer to [...]" since you're using this formulation for the other cases.

Reviewer: I actually suggested the opposite in https://github.com/python/cpython/pull/123544#issuecomment-2323262134, as we use "See" in existing notes like this in `typing.rst` 😄

Reviewer 3: I prefer `See` because it's shorter so I think changing all `Refer to` into `See` would be better. Here, I only suggested what was the most consistent with the new additions :)

Author: Alright, two agreeing opinions are enough to convince me here, updated:)
I decided to stick with "Refer to" initially because it still fits one line and doesn't add significant overhead, but consistency is more important.

>    :class:`Callable` and :class:`~typing.Callable` in type annotations.

Reviewer: Probably don't need to mention the deprecated alias if we're already in the docs for the non-deprecated object. And let's get rid of the link that only takes us to `collections.abc.Callable`, which is where we already are ;)

```suggestion
   :class:`!Callable` in type annotations.
```

Author: I think this bit is relevant enough here - a *lot* of code uses deprecated `typing` aliases, and having both listed here as equivalent helps to understand relationship of these aliases.

Reviewer: Fair enough! Suggestion withdrawn

Author: Sorry, no, agreed - that isn't necessary in `collections.abc` notes. Mentioning both in `typing.rst` guidelines section is sufficient.

   The only reliable way to determine whether an object is :term:`iterable`
   is to call ``iter(obj)``.

.. class:: Collection[T_co](Sized, Iterable[T_co], Container[T_co])

   ABC for sized iterable container classes.

   .. versionadded:: 3.6

.. class:: Iterator[T_co](Iterable[T_co])

   ABC for classes that provide the :meth:`~iterator.__iter__` and
   :meth:`~iterator.__next__` methods.  See also the definition of
   :term:`iterator`.

.. class:: Reversible[T_co](Iterable[T_co])

   ABC for iterable classes that also provide the :meth:`~object.__reversed__`
   method.

   .. versionadded:: 3.6

.. class:: Generator[YieldType_co, SendType_contra, ReturnType_co](Iterator[YieldType_co])

   ABC for :term:`generator` classes that implement the protocol defined in
   :pep:`342` that extends :term:`iterators <iterator>` with the
   :meth:`~generator.send`,
   :meth:`~generator.throw` and :meth:`~generator.close` methods.

   A generator can be annotated by the generic type
   ``Generator[YieldType, SendType, ReturnType]``. For example::

      def echo_round() -> Generator[int, float, str]:
          sent = yield 0
          while sent >= 0:
              sent = yield round(sent)

>           return 'Done'

Reviewer: Most people won't go to the `collections.abc` docs for detailed information on type-hinting. I agree that this information doesn't belong in the docs for the deprecated `typing.Generator` symbol anymore, but I'm not sure this is the right solution. I'd advocate moving the docs on generators and coroutines to an "Annotating generators and coroutines" section towards the top of the `typing` docs, similar to what I did for tuples in https://github.com/python/cpython/commit/968435ddb1c1af9333befb26f7970cded8a5c710, for callables in https://github.com/python/cpython/commit/ca8b55c7f54b38e264056148075a8061a7082013, and for `typing.Type` in https://github.com/python/cpython/commit/3eeb8c89063d5ac22c0b1d26e4ae2fd12c149650

Reviewer: (FWIW, I _do_ think it would be useful to _link_ from the docs for `collections.abc.Generator` to a guide in the typing docs on how to type-hint generators!)

   :meth:`~generator.send`,
   :meth:`~generator.throw` and :meth:`~generator.close` methods.

   Refer to

>    :ref:`annotating generators and coroutines <annotating-generators-and-coroutines>`

Reviewer: I would probably just do the cross-references like this throughout:

```suggestion
   :ref:`annotating-generators-and-coroutines`
```

The only difference is that the word "annotating" is capitalized... but I think that's actually an improvement, as it makes it more obvious that it's the name of a section:

![image](https://github.com/user-attachments/assets/328e5f43-f697-401d-89d6-b7cda66bfb20)

Author: Yeah, agreed - I added this exactly to fight capitalization, but perhaps it's really better to retain it.

>    for details of using this class in type annotations.

Reviewer: ```suggestion
   for details on using this class in type annotations.
```

>    section for details of using this class in type annotations.

Reviewer: ```suggestion
   for details of using this class in type annotations.
```

   ABCs for read-only and mutable :term:`mappings <mapping>`.

   Example use in type annotations::

      def get_position_in_index(word_list: Mapping[str, int], word: str) -> int:
          return word_list[word]

> 

Reviewer: Maybe we should just get rid of this... it doesn't belong in the stub entry for the deprecated alias in `typing`, but it feels out of place here since these docs don't otherwise talk about type annotations at all. And it's too small to deserve its own section in the typing docs, like we're doing for generators/coroutines.

We could also maybe get rid of the "This type may be used as follows:" sections from the stub entries for `typing.List` and `typing.Dict`, as it sort-of implies that it's recommended to use that type, which isn't true. (But for both of those types, I'd keep the "Note that to annotate arguments" warning paragraph.)

Author: And this is a good one, `List` and `Dict` shouldn't suggest such usage.

      Using ``isinstance(gencoro, Coroutine)`` for them will return ``False``.
      Use :func:`inspect.isawaitable` to detect them.

>    The variance and order of type variables
>    correspond to those of :class:`Generator`. Refer to
>    :ref:`annotating generators and coroutines <annotating-generators-and-coroutines>`
>    section for details of using this class in type annotations.

Reviewer: Let's swap these two sentences around:

```suggestion
   Refer to :ref:`annotating generators and coroutines <annotating-generators-and-coroutines>`
   for details of using this class in type annotations.
   The variance and order of type parameters correspond to those of
   :class:`Generator`.
```

      Using ``isinstance(gencoro, Coroutine)`` for them will return ``False``.
      Use :func:`inspect.isawaitable` to detect them.

   Refer to
   :ref:`annotating generators and coroutines <annotating-generators-and-coroutines>`

>    for details of using this class in type annotations.

Reviewer: ```suggestion
   for details on using this class in type annotations.
```

>    The variance and order of type variables correspond to those of

Reviewer: ```suggestion
   The variance and order of type parameters correspond to those of
```

   ABC for :term:`asynchronous generator` classes that implement the protocol
   defined in :pep:`525` and :pep:`492`.

   Refer to
   :ref:`annotating generators and coroutines <annotating-generators-and-coroutines>`

>    section for details of using this class in type annotations.

Reviewer: ```suggestion
   for details of using this class in type annotations.
```

   ABC for :term:`asynchronous generator` classes that implement the protocol
   defined in :pep:`525` and :pep:`492`.

   Refer to
   :ref:`annotating generators and coroutines <annotating-generators-and-coroutines>`

>    for details of using this class in type annotations.

Reviewer: ```suggestion
   for details on using this class in type annotations.
```
